### PR TITLE
Fix misleading Utility message.

### LINKED ISSUE #9713: Utility incorrectly reports that a command does not exist.
screen shot 2015-10-21 at 9 23 39 pm

The command *does* exist, but an `ArgumentException` is being raised by a malformed `Substring`.

```csharp
			catch (Exception e)
			{
				Log.AddChannel("utility", "utility.log");
				Log.Write("utility", "Received args: {0}", args.JoinWith(" "));
				Log.Write("utility", "{0}", e);

			    if (e is ArgumentException)
			        Console.WriteLine("No such command '{0}'", args[0]);
			    else
			    {
                    Console.WriteLine("Error: Utility application crashed. See utility.log for details");
                    throw;
			    }
			}
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27600859-utility-incorrectly-reports-that-a-command-does-not-exist?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by phrohdoh**
Closes #9713.

### CODE REVIEW

**Comment by atlimit8** [comment]
I don't like identifying exception type by checking the message string.
A subclass should be used:
```c#
public class NoSuchCommandException : ArgumentException
{
	public NoSuchCommandException(string command)
		: base("No such command '{0}'", command) { }
}
```
```diff
@@ -57,7 +57,7 @@ static void Main(string[] args)
 			try
 			{
 				if (!actions.ContainsKey(args[0]))
-					throw new ArgumentException();
+					throw new NoSuchCommandException(args[0]);
 
 				var action = actions[args[0]].Key;
 				var validateActionArgs = actions[args[0]].Value;
@@ -78,8 +78,10 @@ static void Main(string[] args)
 				Log.Write("utility", "Received args: {0}", args.JoinWith(" "));
 				Log.Write("utility", "{0}", e);
 
-			    if (e is ArgumentException)
-			        Console.WriteLine("No such command '{0}'", args[0]);
+				if (e is NoSuchCommandException)
+				{
+					Console.WriteLine(e.Message);
+				}
 			    else
 			    {
                     Console.WriteLine("Error: Utility application crashed. See utility.log for details");
```

**Comment by Phrohdoh** [comment]
That's fair.
I'll fix this soon, it is already 12:15am here.

**Comment by Phrohdoh** [comment]
Updated.

**Comment by atlimit8** [comment]
:+1: 
The `command` variable also makes the code a little more readable.

File: OpenRA.Utility/Program.cs [hunk 1]
```diff
@@ -15,6 +15,16 @@
 
 namespace OpenRA.Utility
 {
+	public class NoSuchCommandException : ArgumentException
```
Comment (RoosterDragon) [inline OpenRA.Utility/Program.cs]: This should inherit plain `Exception`.

File: OpenRA.Utility/Program.cs [hunk 2]
```diff
@@ -78,13 +89,13 @@ static void Main(string[] args)
...
 
-			    if (e is ArgumentException)
-			        Console.WriteLine("No such command '{0}'", args[0]);
-			    else
-			    {
-                    Console.WriteLine("Error: Utility application crashed. See utility.log for details");
-                    throw;
-			    }
+				if (e is NoSuchCommandException)
+					Console.WriteLine(e.Message);
```
Comment (RoosterDragon) [inline OpenRA.Utility/Program.cs]: This cleanup enables you to `catch (NoSuchCommandException)` and do the error handling there, which would be better.

File: OpenRA.Utility/Program.cs (hunk 1 above)
Comment (atlimit8) [inline OpenRA.Utility/Program.cs]: How is `NoSuchCommandException` not an `ArgumentException` when it is about an argument.

Comment (RoosterDragon) [inline OpenRA.Utility/Program.cs]: `ArgumentException` defines `ParamName` which hasn't been set at all here, so either that's an oversight or `ArgumentException` doesn't apply in the first place.

Comment (ScottNZ) [inline OpenRA.Utility/Program.cs]: Furthermore, ArgumentExceptions are used for method arguments/parameters, not arguments for commands in OpenRA.Utility

Comment (atlimit8) [inline OpenRA.Utility/Program.cs]: OK.

File: OpenRA.Utility/Program.cs (hunk 2 above)
Comment (atlimit8) [inline OpenRA.Utility/Program.cs]: Then log output would have to be copied or placed in a method.
```c#
 				Log.Write("utility", "Received args: {0}", args.JoinWith(" "));
 				Log.Write("utility", "{0}", e);
```

Comment (Phrohdoh) [inline OpenRA.Utility/Program.cs]: I'd rather keep the amount of changes introduced here to a minimum.

Comment (RoosterDragon) [inline OpenRA.Utility/Program.cs]: Good points.

File: OpenRA.Utility/Program.cs (hunk 1 above)
Comment (Phrohdoh) [inline OpenRA.Utility/Program.cs]: Fixed.

**Comment by Phrohdoh** [comment]
`NoSuchCommandException` now inherits from `Exception`.

**Comment by RoosterDragon** [comment]
:+1:

**Comment by RoosterDragon** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/25ef0d4f364aa966dea233e24d7335df8630009f...3d276d29317d7ac22af2e4e3ece69bc3842f0ca7)

**Pull request merged**